Add AVL implementation

matthiasbeyer opened the pull request:
Hi,

we need a AVL implementation, so I started defining an interface for one. Please review!

@neithernut , @theneikos , @manuelmessner

matthiasbeyer committed:
Added avl.{c,h} files with first definition of the interface

matthiasbeyer committed:
Added implementation of avl_alloc(), which is just an alias for convenience

matthiasbeyer committed:
Added comperator func ptr arg to avl_add()

And adapted avl_alloc() appropriately.

matthiasbeyer committed:
Added infrastructure in avl_add() function

matthiasbeyer committed:
Fixup of the parameters of the find() and del() functions

matthiasbeyer committed:
Added helper function for finding elements

matthiasbeyer committed:
Added infrastructure implementation for avl_del() function

matthiasbeyer committed:
Added more comment to the helper function

matthiasbeyer committed:
Added implementation of the avl_find() function

matthiasbeyer committed:
Fixup for the initial import (argument types)

matthiasbeyer committed:
Added basic infrastructure implementation of the avl_unlink() function

matthiasbeyer committed:
Fix: We should name the data variable "data", not "d"

Suggested-by: Marcel Mueller <com2040@gmail.com>

matthiasbeyer committed:
Added documentation for the AVL tree node data type

Suggested-by: Marcel Mueller <com2040@gmail.com>

TheNeikos commented on src/libreset/avl.h:
```diff
@@ -0,0 +1,78 @@
+#define __AVL_H__
+
+struct avl_el {
+    void*   d;
```
If this is data, please rename it as such, or documentation

matthiasbeyer committed:
Added simple implementation for avl_destroy()

TheNeikos commented on src/libreset/avl.h:
```diff
@@ -0,0 +1,84 @@
+avl_add(
+        struct avl_el* root, //!< The root of the avl tree where to insert
+        int (*cmp)(void* const, void* const), //!< the comperator function
+        void* const d //!< The data element
```
I just notived that there is also this one here that would need changing if we want to stay consistent.

matthiasbeyer replied:
I guess this diff is already outdated. I just fixed this.

neithernut commented on src/libreset/avl.h:
```diff
@@ -0,0 +1,84 @@
+ * @return 1 on success, else a negative error code
+ */
+int
+avl_del(
```
We may want to return a pointer to the data here, for optional further usage.
If the element was not found, NULL may be returned.

neithernut commented:
The comparator's return value has to be defines. Do we return [-1,0,1](preferable) or [0,1]?
Hint: this is a perfect example of when to use a typedef for  a function type.

Note: we will use the hash as a key inside the AVL tree, so supplementing a function isn't neccessary anyway.

matthiasbeyer committed:
Fix: Return ptr to the deleted element on avl_del()

matthiasbeyer committed:
Fix: Added missing additional parameter for comparator function of avl_add

matthiasbeyer commented:
So you mean we do not use cmp-functions? So should I remove the `cmp` parameter from the functions
- `avl_add`
- `avl_del`
- `avl_find`
- `avl_unlink`

?

neithernut commented:
Yep

matthiasbeyer commented:
I'll do this now, but we need some helper functions anyways, for the `avl_del()`, we need something which determines which element gets removed.

matthiasbeyer committed:
Remove the `cmp` argument for the `avl_add()` function

matthiasbeyer committed:
Removed `cmp` argument for `avl_{del,find,unlink}`

This patch removes the `cmp` argument for the functions

    * avl_del
    * avl_find
    * avl_unlink

and replaces it with a `pred` (predicate) argument, as we somehow need a
selector for the functions.

matthiasbeyer commented:
Anyways, this needs further patching, as the actual algorithms are not implemented.

Also, this needs thread-safeness.

matthiasbeyer committed:
Added hash member for the avl_el type

matthiasbeyer committed:
Added hash getter function for the avl_el type

matthiasbeyer committed:
Added import for hash.h header

matthiasbeyer committed:
Added implementation for find_element_with_parent() helper function

matthiasbeyer committed:
Added height member (including getter) for the avl_el type

matthiasbeyer committed:
Added single and double rotation helpers

matthiasbeyer committed:
Implemented insert algorithm

matthiasbeyer committed:
Use new insert() function for inserting into tree

matthiasbeyer commented on src/libreset/util/macros.h:
```diff
@@ -0,0 +1,14 @@
+#define MIN(x,y)                    \
+       ({ __typeof__ (x) _x = (x);  \
+          __typeof__ (y) _y = (y);  \
+          _x < _y ? _x : _y; })
```
These should be typesafe because we actually use variables in here.

We use a block `({...})`, so the temporary variables get out of scope after the macros are used.

matthiasbeyer commented:
I guess I need help on this. I guess we should do this together.

Remaining:
- Implement deleting
- Make threadsafe (I have no experience on this at all)

Do we need algorithms for merging two AVLs? If yes, this would be another point.

matthiasbeyer committed:
Shortened avl_height() implementation

matthiasbeyer commented on src/libreset/avl.h:
```diff
@@ -0,0 +1,137 @@
+ * The element of the delete operation is returned, so we do not have a dangling
+ * pointer laying around.
+ *
+ * @return the element on success, else NULL
```
I think we should better return the new root node, instead of the element. This sounds much more obvious, as with the rotation, the root can change!

neithernut replied:
Yep.

matthiasbeyer committed:
Added helper function to return higher children node of a root node

matthiasbeyer committed:
Added helper for getting the height difference between two subtrees

matthiasbeyer committed:
Added helper for recomputing the height for a subtree

matthiasbeyer committed:
Added header file documentation

matthiasbeyer committed:
Added documentation group for attribute-macros

matthiasbeyer commented:
This set gets huge. I think we should refactor the whole set after the implementation is ready, to shorten the history a bit, if possible.

And if you want, of course, because I don't think it is that urgent.

---

Edit:

Maybe we should re-roll this and add a `devel` and a `devel-avl` branch, whereas the latter is for developing the AVL implementation?
This way, there would be the possibility to get the utils which are added by this set (macros.h, element.h, hash.h) into the `devel` branch beforehand.

@neithernut , @TheNeikos thoughts?

TheNeikos commented:
Yes to the second one.

I am not so sure about the benefits of splitting up the branches other than a different view.

matthiasbeyer commented:
Well, of course it is a benefit mainly for the view.

---

Besides: You are (all) welcome to submit PRs onto this branch! When doing a PR, you can select where this PR should be merged, and you can simply define this branch as merge-point. So we can collaborate on one branch.
I would then do a `--ff` merge (offline, github can detect this), as this branch will be rebased (due to #10 ) anyway!

matthiasbeyer commented:
Some commits can be removed after #24 , #25 and #26  are merged.

matthiasbeyer committed:
Added helper for finding leftmost node in a subtree

matthiasbeyer committed:
Implemented deleting a node (known BUG included)

matthiasbeyer committed:
Added @todo in avl_del()

neithernut commented:
Regarding merging and splitting functions: for now, construction/initialisation, destruction insertion and deletion should be enough. All the other things may go into another PR.

matthiasbeyer commented:
:+1:

matthiasbeyer committed:
Added avl type which contains the avl tree

matthiasbeyer committed:
Rewrote interface for avl type

Added `struct avl` type, which contains the tree.

This required a re-write of the interface definition for the
`struct avl` type.

matthiasbeyer committed:
Adapted implementation of avl_alloc() to new interface

matthiasbeyer committed:
Adapted implementation of avl_destroy() to new interface

Introducing a new helper function for destroying a subtree.

matthiasbeyer committed:
Adapted avl_add() to new interface definition

Modifying insert() appropriately.

matthiasbeyer committed:
Adapted current avl_del() implementation to interface definition

matthiasbeyer committed:
Adapted avl_find() implementation to new interface definition

matthiasbeyer committed:
Adapted avl_unlink() implementation to new interface definition

matthiasbeyer committed:
Added typdef for predicate function type

matthiasbeyer committed:
Removed avl_unlink()

matthiasbeyer commented:
Updated to latest master, including removing commits which were added to master already by other PRs already (mainly `src/libreset/hash.h`).

And removed `src/libreset/element.h` from the history by reverting and deleting appropriate patches.

neithernut commented on src/libreset/avl.c:
```diff
@@ -0,0 +1,421 @@
+    int f = find_element_with_parent(avl->root, pred, etc, &parent, &found);
+
+    if (f) {
+
```
In such cases, I prefer an early return. The program flow is much clrearer this way imho and you save some intendations.

neithernut commented on src/libreset/avl.c:
```diff
@@ -0,0 +1,421 @@
+
+    /*
+     * @todo rebalancing, with rotations if neccessary.
+     */
```
The rebalancing method will be called by the higher level methods of the set.

matthiasbeyer replied:
So there will be no rebalancing in this function?

neithernut replied:
nope

neithernut commented on src/libreset/avl.c:
```diff
@@ -0,0 +1,421 @@
+static struct avl_el*
+insert(struct avl_el* new, struct avl_el** root)
+{
+    signed int cmpres = new->hash == root->hash ? 0 : (new->hash > root->hash ? -1 : 1);
```
Shouldn't it be (*root)->hash?
Does this even compile?
If the computation of cmpres invlolves the trinary operator, it's pointless to not put the contidion right into the expression at the "if" itself anyway.

matthiasbeyer replied:
Thank you for reviewing this carefully! The rebase messed this up. Will add a fixup patch, of course.

neithernut commented on src/libreset/avl.c:
```diff
@@ -0,0 +1,421 @@
+
+    if (*root == NULL) {
+        return new;
+    } else if (cmpres == 1) {
```
This chunk is malicious, or malformed, I think.
You should check for *root==NULL first (returning if NULL) and determine cmpres _afterwards_
Has the advantage of not needing the else

matthiasbeyer committed:
Removed indentation level by negating condition

matthiasbeyer committed:
We can simply open a new if here

matthiasbeyer commented:
Will rebase (and merge the fixup commits in) after #10 is merged.

matthiasbeyer commented:
I rewrote all commits to include a pre-commit hook run (run astyle after each commit). It was fiddeling around, but I guess it worked.

matthiasbeyer committed:
Expanded doubled ternary operator

matthiasbeyer commented:
Rebased and rewrote history due to #31 .

matthiasbeyer committed:
Rewrote (internal) find_element_with_parent() helper

This patch rewrites the (internal) find_element_with_parent() helper
function.

I'm not even sure the previous version worked, but I assume this one to
work (lack of testing by now).

The approach of this:

    1.  Check if the passed `struct avl_el` exists, return 0 if not, do
        not touch anything.

    2.  If the root node is the node we search:
        2.1.    Set the `*parent_dest` to NULL, as we assume this is
                the first recursion step

        2.2.    If the `found_dest` exists, set it.

        2.3.    return 1, we found it

    3. Check the left and the right subtree by doing a recursion step.

        Found:  If the recursion step has found the node we look for, set
                `*parent_dest` accordingly and only if it is not set
                yet.
                This prevents further (upper) recursion steps from
                re-setting it, which would result the root node of the
                tree we operate on to be saved as the parent node, which
                is wrong.

                return 1 afterwards, we found it.

        Not:    The node we look for is not in the tree. Ensure the
                `*parent_node` is NULL and return 0, nothing found.

matthiasbeyer committed:
Add extensive documentation for find_element_with_parent()

matthiasbeyer committed:
We actually should use the typedef for the predicate function type here

matthiasbeyer commented on src/libreset/avl.c:
```diff
@@ -0,0 +1,486 @@
+/**
+ * Recompute the height for a subtree
+ *
+ * @invariant \f$ -1 \leqslant H(n) \leqslant +1 \f$, Where \f$ H(n) \f$ is the
```
Note: Requires "amssymb" latex package to be included for the doxygen documentation

matthiasbeyer committed:
Use \leq instead of \leqslant, so we do not need amssymb

matthiasbeyer commented:
@neithernut please review and tell me what needs to be done in this PR. @manuelmessner and @TheNeikos you're welcome, too!

matthiasbeyer commented:
With #35 , things get more complicated with the implementation this PR ships. Things have to be reworked. I would like to see some PRs onto this branch transforming it to the use of the linkedlist implementation ( #35 ).

matthiasbeyer committed:
Use linkedlist implementation for element chaining

matthiasbeyer committed:
Updated interface: on avl_add(), we want the hash for the element, too

matthiasbeyer committed:
Adapted avl_add() to the new interface definition, removing the old insertion algorithm

matthiasbeyer committed:
Added algorithm to find node by hash

matthiasbeyer committed:
Added interface for internal helper function for inserting avl_el object into tree

matthiasbeyer committed:
Added new implementation for internal insert() helper, updated documentation

neithernut commented:
Wait... you mean reimplement the insertion function you already wrote?

matthiasbeyer commented:
Not entirely.

The insertion algorithm is implemented for an element for which the hash is already in the tree. It is unimplemented for a node where the hash is not in the tree already.

Edit: I just can't wrap my head around the whole rotation foo. I know how it works, but I don't know when to  rotate left and when right, leave alone single and double rotations!

---

Edit: I can try, of course...

matthiasbeyer committed:
Added implementation for insert_node_into_tree()

neithernut commented:
If you have real problems, strip the implementation and push the interface together with stubs.
We may then open an issue for each function not implemented and assign functions to people.
Rotations have, of course, be implemented first.

matthiasbeyer commented:
Rotations are already implemented!

matthiasbeyer commented:
Rebased and squashed fixup commits.

neithernut commented:
To give the travic CI integration test some meaning, the source files have to be included in the CMake setup.

matthiasbeyer commented:
You are the expert on this, please add an appropriate PR onto this branch about this.

Isn't this covered by the `add_subdirectory(src)` line in the `CMakeLists.txt` file?

neithernut commented:
Nope, there is a second CMakeLists.txt in src. If you open that one for editing, it should be obvious to you what you have to do.

matthiasbeyer committed:
Added avl.c to build inputs

matthiasbeyer commented:
@neithernut How to add an include path? The build fails because the appropriate header wasn't found!

(`/home/travis/build/matthiasbeyer/libreset/src/libreset/avl.c:1:26: fatal error: libreset/avl.h: No such file or directory`)

matthiasbeyer committed:
Added include path in CMakeLists.txt file

matthiasbeyer committed:
Fix: typedef in hash.h

matthiasbeyer committed:
Added missing include in avl.h : util/attributes.h

matthiasbeyer committed:
Fix: we do not need to set the data element

Fix: we do not need to set the data element (which does not exist
anyways) to NULL here.

matthiasbeyer committed:
Fix: Change order of includes, so we do not override some stdlib stuff with the macros which get included

matthiasbeyer committed:
Removed commata to fix compilation error

matthiasbeyer committed:
Removed ignored function attribute

neithernut commented:
But you still have to add the source files to `SOURCE_FILES` (line 12 in src/CMakeLists.txt)

matthiasbeyer commented:
@neithernut It lists the `avl.c` file, doesn't it? :-P

matthiasbeyer commented:
For documentation: due to #46 , the code comment converting commits (cccc - lol!) can and will be removed.

neithernut commented:
Ahh, I looked at an old commit.

matthiasbeyer commented:
@neithernut : Would you have a look at [the travis build](https://travis-ci.org/matthiasbeyer/libreset/builds/34787455), it still complains about ISO-C.

neithernut commented:
The warning message is clear: ISO-C forbids blocks in expressions.
Either the MAX macro has to be simplyfied to something likee:

``` C
#define MAX(a,b) ((a)>(b)?(a):(b))
```

or we do _not_ write ISO-C.

matthiasbeyer commented:
The idea of the macro is that it does a type safe comparison. You cannot do 

``` C
int a = 1;
double b = 2.0;
int c = MAX(a, b);
```

without getting a compiler warning about it, if we use the macro as it is defined. The simplification wouldn't be typesafe...

matthiasbeyer committed:
Added actually missing type check in MAX/MIN macros

matthiasbeyer commented:
I would ACK an appropriate patch to the build environment for fixing this rather as to a patch changing the macros so they do not contain blocks.

Macros with type safety are much more of an improvement than ISO-C vs. non-ISO-C.

---

(At least: Does clang support the _full_ ISO C standard? Because maybe we want to compile the library with clang later...)

matthiasbeyer committed:
Fix: Inline documentation shouldn't be included in the doxygen documentation

neithernut commented:
clang is supposed to be gcc-compatible, hence it should also support braces as part of an expression. But that's not my point here. Not conforming to the ISO-C standart is not acceptable imho.
The entire typesafety argument may be an issue we might consider neccessary to prevent, however, if this means _not_ doing ISO-C the solution is flawed.
I'd suggest using explicit casts without resorting to assignments. I'll push my sollution and add a PR.

neithernut commented:
I had a closer look at it and actually, I don't get in what way this does type checking, anyway.

neithernut commented:
Please have a look at PR #49.

neithernut commented:
Also, _please_ divide this branch into interface and implementation(s) and merge it already, so we can properly assign functions implementation to people.

matthiasbeyer commented:
I guess we should just restart this PR. The history is indeed a mess.

matthiasbeyer commented:
I'll close this now. That branch is a good example for a real fucked up history.
